statick_tool/plugins/tool/spotbugs_tool_plugin.py

```diff
@@ -0,0 +1,119 @@
+
+from statick_tool.issue import Issue
+from statick_tool.tool_plugin import ToolPlugin
+
+
+class SpotbugsToolPlugin(ToolPlugin):
+    """Apply spotbugs tool and gather results."""
+
+    def get_name(self):
+        """Get name of tool."""
+        return "spotbugs"
+
+    def get_tool_dependencies(self):
+        """Get a list of tools that must run before this one."""
+        return ["make"]
+
+    def scan(self, package, level):
+        """Run tool and gather output."""
+        # Sanity check - make sure mvn exists
+        if not self.command_exists('mvn'):
+            print("Couldn't find 'mvn' command, can't run Spotbugs Maven integration")
+            # Return [] instead of None because this isn't an error per se,
+            # and we don't want to fail because a tool is missing
+            return []
+
+        flags = ["-Dspotbugs.effort=Max", "-Dspotbugs.threshold=Low",
+                 "-Dspotbugs.xmlOutput=true"]
+        flags += self.get_user_flags(level)
+
+        include_file = self.plugin_context.config.get_tool_config(self.get_name(),
+                                                                  level, "include")
+        exclude_file = self.plugin_context.config.get_tool_config(self.get_name(),
+                                                                  level, "exclude")
+        if include_file is not None:
+            flags += ["-Dspotbugs.includeFilterFile={}".format(self.plugin_context
+                                                               .resources.get_file(include_file))]
+
+        if exclude_file is not None:
+            flags += ["-Dspotbugs.excludeFilterFile={}".format(self.plugin_context
+                                                               .resources.get_file(exclude_file))]
+
+        issues = []
+        with open(self.get_name() + ".log", "w") as f:
+            for pom in package['top_poms']:
+                try:
+                    # The spotbugs:spotbugs-maven-plugin split is auto-concatenated
+                    output = subprocess.check_output(["mvn", "com.github.spotbugs:"
+                                                      "spotbugs-maven-plugin:spotbugs"] +
+                                                     flags,
+                                                     cwd=os.path.dirname(pom),
+                                                     stderr=subprocess.STDOUT,
+                                                     universal_newlines=True)
+                except subprocess.CalledProcessError as ex:
+                    output = ex.output
+                    f.write(output)
+                    print("spotbugs failed! Returncode = {}".
+                          format(str(ex.returncode)))
+                    print("{}".format(ex.output))
+                    return None
+
+                except OSError as ex:
+                    print("Couldn't find maven! (%s)" % (ex))
+                    return None
```

Comment:
Earlier in this method you are returning `[]` instead of `None` when the `mvn` command is not found. Is there a reason to have a different return value?

Do you want to be checking again that this command exists? I think you should leave this extra check/error handling, since the first check may change in the future.

Reply:
I agree that they're inconsistent, but I'll bounce part of the question back to you - do you think a missing tool should be considered an error? I'm inclined to say "no" but I could see arguments both ways. I'd like to leave this check in just for safety's sake - I want to make sure that the subprocess call can't throw an unhandled exception, and I have seen times in the past where subprocess somehow couldn't find a tool which command_exists found (still trying to make a reproducible test case for that...)

Reply:
For considering a missing tool to be an error, I can see arguments both ways.

If you are expecting this tool to tell you when some constraints have been violated, and there is a specific tool to determine whether a set of constraints were violated, then a missing tool means you will not have any assurances that you are within desired constraints. That seems like something this tool should alert the user about. Right now, the only way I can think to create that alert is to return an error.

On the other hand, some users might want to use most of the tools for an existing level and are fine tracking upstream (so they don't have to create a custom config and merge upstream changes into it) but only installing the tools they care about. Or, some users might not be able to configure their tools for use in third party CI tools (such as Travis).

I'm not sure what all the use cases are for this tool. Right now I think it's only me and you :) I'm fine following your lead if you want to allow Statick to return success even when there are missing tools. In the future, I might go and add a "global" flag to Statick to allow the user to select how they want Statick to treat missing tools. Something like a `set_error_on_missing_tool` variable.